source/pervasive/invariants.rs

```diff
@@ -0,0 +1,80 @@
+#[allow(unused_imports)] use builtin::*;
+#[allow(unused_imports)] use crate::pervasive::*;
+
+// TODO right now this file contains the minimal type definitions needed for rust_verify
+// but we need to make a library to make the Invariant type actually usable
+//  * constructors, destructors
+//  * utility for creating unique namespaces
+
+#[proof]
+#[verifier(external_body)]
+pub struct Invariant<V> {
+    dummy: std::marker::PhantomData<V>,
+}
+
+// note the names of `inv` and `namespace` are harcoded.
+// TODO right now they are only translated if the user refers to them in their code
+// which can cause a crash otherwise
+
+#[proof]
+impl<V> Invariant<V> {
+    #[verifier(pub_abstract)]
+    #[spec]
+    pub fn inv(&self, _v: V) -> bool {
+        arbitrary()
+    }
+
+    // If you want to open two invariants I and J at the same time,
+    // you need to show that I.namespace() != J.namespace()
+    // The namespace can be declared upon allocation of the invariant.
+
+    #[verifier(pub_abstract)]
+    #[spec]
+    pub fn namespace(&self) -> int {
+        arbitrary()
+    }
+}
+
+#[proof]
+pub struct InvariantBlockGuard;
+
+// NOTE: These 2 methods are removed in the conversion to VIR; they are only used
+// for encoding and borrow-checking.
+// In the VIR these are all replaced by the OpenInvariant block.
+// This means that the bodies, preconditions, and even their modes are not important.
+//
+// An example usage of the macro is like
+//
+//   i: Invariant<X>
+//
+//   open_invariant!(&i => inner => {
+//      { modify `inner` here }
+//   });
+//
+//  where `inner` will have type `X`.
+//
+//  The purpose of the `guard` object, used below, is to ensure the borrow on `i` will
+//  last the entire block.
+
+#[verifier(external_body)]
+pub fn open_invariant_begin<'a, V>(_inv: &'a Invariant<V>) -> (&'a InvariantBlockGuard, V) {
+    requires([false]);
+    unimplemented!();
+}
+
+#[verifier(external_body)]
+pub fn open_invariant_end<V>(_guard: &InvariantBlockGuard, _v: V) {
+    requires([false]);
+    unimplemented!();
+}
+
+#[macro_export]
+macro_rules! open_invariant {
+    ($eexpr:expr => $iident:ident => $bblock:block) => {
+        #[verifier(invariant_block)] {
+            #[allow(unused_mut)] let (guard, mut $iident) = crate::open_invariant_begin($eexpr);
+            $bblock
+            crate::open_invariant_end(guard, $iident);
```

Comment:
Why aren't these `crate::pervasive::invariants::open_invariant_begin` / `crate::pervasive::invariants::open_invariant_end` ? Is it possible that you are relying on this behaviour https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3fe821824832fae9cec28092d6d2c05e in the tests and examples? I.e. importing `invariants::*` at the crate root makes them appear to be at the crate root.

Reply:
are you saying that another module importing this one won't work unless they import `invariants::*`? I thought macros were supposed to do the right thing with identifier lookups. Else how could the macro possibly work across crates?

(i don't know what i'm talking about, please set me straight)

Reply:
Those functions' canonical paths are `crate::pervasive::invariants::open_invariant_begin` and `crate::pervasive::invariants::open_invariant_end` I believe, and _I believe_ this macro only works if you're importing `invariants::*` **and you are at the crate root** (which is the case for all examples and tests, I think). Try the canonical path instead? It should just work.

> Else how could the macro possibly work across crates?

The way `pervasive` is included is a hack currently. It's imported as if it was source in the current crate's `crate::pervasive` module. If you were writing a macro to be used outside the crate, you'd have to use `crate_name::canonical::path` instead, I think. The issue is that macros are expanded before resolution, so they cannot perform identifier lookups. Ping me on Slack if this doesn't make sense yet.